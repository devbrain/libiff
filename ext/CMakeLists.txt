include(FetchContent)

set(FETCHCONTENT_QUIET off)

# Save current warning flags
set(SAVED_CMAKE_CXX_FLAGS ${CMAKE_CXX_FLAGS})
set(SAVED_CMAKE_C_FLAGS ${CMAKE_C_FLAGS})

# Disable warnings for third-party libraries
if(CMAKE_CXX_COMPILER_ID MATCHES "GNU|Clang")
    set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -w")
    set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -w")
elseif(CMAKE_CXX_COMPILER_ID STREQUAL "MSVC")
    set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} /W0")
    set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} /W0")
endif()

# Also set properties to suppress warnings in a more modern way
add_compile_options(
    $<$<CXX_COMPILER_ID:GNU,Clang>:-w>
    $<$<CXX_COMPILER_ID:MSVC>:/W0>
)

add_subdirectory(doctest)

# Restore original flags (though subdirectories won't inherit the change back)
set(CMAKE_CXX_FLAGS ${SAVED_CMAKE_CXX_FLAGS})
set(CMAKE_C_FLAGS ${SAVED_CMAKE_C_FLAGS})


