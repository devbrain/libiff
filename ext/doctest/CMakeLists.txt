cmake_minimum_required(VERSION 3.20)

# Try to find doctest or fetch it
find_package(doctest QUIET)

if (NOT doctest_FOUND)
    message(STATUS "doctest not found, fetching from GitHub...")

    include(FetchContent)
    set(FETCHCONTENT_QUIET OFF)

    # Allow FetchContent_Populate to be used
    if (POLICY CMP0169)
        cmake_policy(SET CMP0169 OLD)
    endif ()

    set(DOCTEST_WITH_TESTS OFF CACHE INTERNAL "")
    set(DOCTEST_WITH_MAIN_IN_STATIC_LIB OFF CACHE INTERNAL "")
    set(DOCTEST_NO_INSTALL ON CACHE INTERNAL "")

    FetchContent_Declare(
            doctest
            GIT_REPOSITORY https://github.com/onqtam/doctest.git
            GIT_TAG v2.4.12
            GIT_PROGRESS TRUE
    )

    FetchContent_MakeAvailable(doctest)
endif ()

# Export doctest for use in parent project
if (doctest_FOUND OR TARGET doctest::doctest)
    message(STATUS "doctest available for testing")
endif()