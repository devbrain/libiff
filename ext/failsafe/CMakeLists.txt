# First try to find failsafe locally
find_package(failsafe QUIET)

if(NOT failsafe_FOUND)
    message(STATUS "failsafe not found locally, fetching from GitHub...")
    
    # Disable failsafe tests and examples
    set(FAILSAFE_BUILD_TESTS OFF CACHE INTERNAL "")
    set(FAILSAFE_BUILD_EXAMPLES OFF CACHE INTERNAL "")
    
    fetchcontent_declare(
                FAILSAFE
                GIT_REPOSITORY "https://github.com/devbrain/failsafe.git"
                GIT_TAG master
                GIT_PROGRESS TRUE
                UPDATE_COMMAND ""
                OVERRIDE_FIND_PACKAGE
        )
    fetchcontent_makeavailable(FAILSAFE)
    add_library(thirdparty::failsafe ALIAS failsafe)
else()
    message(STATUS "Found failsafe locally")
    if(NOT TARGET thirdparty::failsafe)
        add_library(thirdparty::failsafe ALIAS failsafe::failsafe)
    endif()
endif()
