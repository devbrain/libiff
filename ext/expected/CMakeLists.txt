# Check if std::expected is available
include(CheckCXXSourceCompiles)
set(CMAKE_REQUIRED_FLAGS "${CMAKE_CXX_FLAGS} -std=c++23")
check_cxx_source_compiles("
    #include <expected>
    int main() {
        std::expected<int, char> e{42};
        return e.value();
    }
" HAVE_STD_EXPECTED)

if(HAVE_STD_EXPECTED)
    message(STATUS "Using std::expected from C++23 standard library")
else()
    message(STATUS "std::expected not available, using tl::expected fallback")
endif()

if(NOT HAVE_STD_EXPECTED)
    # First try to find tl::expected if it's already installed
    find_package(tl-expected QUIET)
    
    if(tl-expected_FOUND)
        message(STATUS "Found installed tl::expected version ${tl-expected_VERSION}")
        add_library(thirdparty::expected ALIAS tl::expected)
    else()
        message(STATUS "tl::expected not found, fetching from GitHub...")
        # Fetch tl::expected only if std::expected is not available and tl::expected is not installed
        set(EXPECTED_BUILD_PACKAGE OFF CACHE INTERNAL "")
        set(EXPECTED_BUILD_TESTS OFF CACHE INTERNAL "")
        fetchcontent_declare(
                EXPECTED
                GIT_REPOSITORY "https://github.com/TartanLlama/expected.git"
                GIT_TAG v1.2.0
                GIT_PROGRESS TRUE
                UPDATE_COMMAND ""
        )
        fetchcontent_makeavailable(EXPECTED)
        add_library(thirdparty::expected ALIAS expected)
    endif()
else()
    # Create an interface library that uses std::expected
    add_library(thirdparty_expected_std INTERFACE)
    add_library(thirdparty::expected ALIAS thirdparty_expected_std)
endif()